From f100c47eff6a24a9594702dc72037773f7dd8911 Mon Sep 17 00:00:00 2001
From: Eyal Reizer <eyalr@ti.com>
Date: Tue, 29 May 2012 14:45:07 +0300
Subject: [PATCH] wl12xx: detect between wl1271 PG 3.0 and lower

* Detect between wl1271 PG version 3.0 and lower for using the right
  firmware with R5 based releases
* Do not allow activating multi-role on PG 2.0(and lower)

Upstream-Status: Pending

Signed-off-by: Eyal Reizer <eyalr@ti.com>
---
 drivers/net/wireless/wl12xx/main.c |   22 ++++++++++++++++++++--
 1 files changed, 20 insertions(+), 2 deletions(-)

diff --git a/drivers/net/wireless/wl12xx/main.c b/drivers/net/wireless/wl12xx/main.c
index 3f05874..db50bf0 100644
--- a/drivers/net/wireless/wl12xx/main.c
+++ b/drivers/net/wireless/wl12xx/main.c
@@ -51,6 +51,8 @@
 #include "scan.h"
 #include "version.h"

+#define WL1271_PG_VERSION_2 2
+#define WL127X_PG20_FW_NAME_SINGLE	"ti-connectivity/wl127x-fw-4-pg2-0-sr.bin"
 #define WL1271_BOOT_RETRIES 3

 static struct conf_drv_settings default_conf = {
@@ -1081,8 +1083,12 @@ static int wl12xx_fetch_firmware(struct wl1271 *wl, bool plt)
			fw_type = WL12XX_FW_TYPE_NORMAL;
			if (wl->chip.id == CHIP_ID_1283_PG20)
				fw_name = WL128X_FW_NAME_SINGLE;
-			else
-				fw_name = WL127X_FW_NAME_SINGLE;
+			else {
+				if (wl->hw_pg_ver > WL1271_PG_VERSION_2)
+					fw_name = WL127X_FW_NAME_SINGLE;
+				else
+					fw_name = WL127X_PG20_FW_NAME_SINGLE;
+			}
		}
	}

@@ -2494,6 +2500,18 @@ static int wl1271_op_add_interface(struct ieee80211_hw *hw,
		goto out;
	}

+	/*
+	 * Only WL127x PG 3.0 supports multi-role. Do not allow bringing up more
+	 * than one interface at a time in-case of using an older version.
+	 */
+	if ((wl->chip.id == CHIP_ID_1271_PG20) && (wl->hw_pg_ver <= WL1271_PG_VERSION_2)) {
+		open_count = ieee80211_get_open_count(hw, vif);
+		if (open_count) {
+			wl1271_error("WL127x PG %d.0 doesn't support multi-role",wl->hw_pg_ver + 1);
+			ret = -EBUSY;
+			goto out;
+		}
+	}

	ret = wl12xx_init_vif_data(wl, vif);
	if (ret < 0)
--
1.7.0.4
